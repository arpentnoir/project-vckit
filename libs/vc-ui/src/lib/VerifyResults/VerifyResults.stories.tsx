import type { ComponentMeta, ComponentStory } from '@storybook/react';
import { validResults, invalidResults } from '../fixtures';
import { VerifyResults } from './VerifyResults';

const Story: ComponentMeta<typeof VerifyResults> = {
  component: VerifyResults,
  title: 'VerifyResults',
};
export default Story;

const Template: ComponentStory<typeof VerifyResults> = (args) => (
  <VerifyResults {...args} />
);

export const Valid = Template.bind({});

Valid.args = validResults;

export const Invalid = Template.bind({});
Invalid.args = invalidResults;
