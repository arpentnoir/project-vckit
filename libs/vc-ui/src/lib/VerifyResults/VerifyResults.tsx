import CheckCircleOutlineOutlined from '@mui/icons-material/CheckCircleOutlineOutlined';
import CancelOutlinedIcon from '@mui/icons-material/CancelOutlined';
import { Box, Stack } from '@mui/material';
import React from 'react';
import { Text } from '../Text';
import { Issuer, VerifyResult } from '../VerifyViewer';

export interface IVerificationResults {
  issuer: Issuer;
  result: string[];
}

export type CHECK_TYPE = 'proof' | 'status';

export type CheckObject = {
  type: CHECK_TYPE;
  validText: string;
  invalidText: string;
};

export const CHECKS_INFO: CheckObject[] = [
  {
    type: 'proof',
    validText: 'Document has not been tampered with',
    invalidText: 'Document has been tampered with',
  },
  {
    type: 'status',
    validText: 'Document has not been revoked',
    invalidText: 'Document has been revoked',
  },
];

export interface IValid {
  issuer: Issuer;
  checks: CHECK_TYPE[];
  errors: CHECK_TYPE[];
}

export const getCheckStatus = (check: CHECK_TYPE, errors: CHECK_TYPE[]) => {
  const item = CHECKS_INFO.filter((item) => item['type'] === check);

  if (item.length !== 1) {
    return null;
  }

  return errors.includes(item[0].type)
    ? {
        isValid: false,
        text: item[0].invalidText,
      }
    : {
        isValid: true,
        text: item[0].validText,
      };
};

export const displayCheckResults = (
  checks: CHECK_TYPE[],
  errors: CHECK_TYPE[]
) => {
  return (
    <>
      {checks.map((check) => {
        const checkObject = getCheckStatus(check, errors);

        return (
          checkObject && (
            <Box
              alignItems="center"
              display="flex"
              key={check}
              data-testid={check}
            >
              {checkObject.isValid ? (
                <CheckCircleOutlineOutlined color="success" />
              ) : (
                <CancelOutlinedIcon color="error" />
              )}

              <Text
                variant="body1"
                fontWeight={400}
                fontStyle="italic"
                display="inline"
                paddingLeft="12px"
              >
                {checkObject.text}
              </Text>
            </Box>
          )
        );
      })}
    </>
  );
};

export const Results = ({ issuer, checks, errors }: IValid) => {
  const hasErrors = errors.length !== 0;
  const statusText = hasErrors ? 'Invalid' : 'Valid';

  return (
    <React.Fragment>
      <Box
        bgcolor={hasErrors ? '#dc3545' : '#28a745'}
        borderRadius="3px"
        display="inline-flex"
        data-testid="valid-box"
        marginBottom="36px"
      >
        <Text
          color="white"
          fontWeight="bold"
          variant="h5"
          padding="8px 16px"
          aria-label={`Document is ${statusText}`}
        >
          {statusText}
        </Text>
      </Box>
      <Stack
        direction={{ xs: 'column', md: 'row' }}
        spacing={{ xs: '16px', md: '64px' }}
      >
        {/* Only show issuer if VC is valid */}
        {!hasErrors && (
          <Box>
            <Text variant="h6">Issued by:</Text>
            <Text
              fontWeight="bold"
              variant="h6"
              style={{ wordWrap: 'break-word' }}
            >
              {issuer.id}
            </Text>
          </Box>
        )}

        {displayCheckResults(checks, errors)}
      </Stack>
    </React.Fragment>
  );
};

export const VerifyResults = ({ issuer, checks, errors }: VerifyResult) => {
  return (
    <Results
      issuer={issuer}
      checks={checks as CHECK_TYPE[]}
      errors={errors as CHECK_TYPE[]}
    />
  );
};
