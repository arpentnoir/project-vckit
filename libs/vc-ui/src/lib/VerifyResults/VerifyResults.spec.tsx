import { render } from '@testing-library/react';
import { invalidResults, validResults } from '../fixtures';
import { VerifyResults } from './VerifyResults';

describe('VerifyResults', () => {
  it('should render correctly', () => {
    const { baseElement } = render(
      <VerifyResults {...(validResults as any)} />
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should display correct checks for a valid document', () => {
    const { getByText } = render(<VerifyResults {...validResults} />);

    getByText('Valid');
    getByText('Issued by:');
    getByText('did:key:123');
    getByText('Document has not been tampered with');
    getByText('Document has not been revoked');
  });

  it('should display correct checks for an invalid document', () => {
    const { getByText } = render(<VerifyResults {...invalidResults} />);

    getByText('Invalid');
    getByText('Document has been tampered with');
  });
});
