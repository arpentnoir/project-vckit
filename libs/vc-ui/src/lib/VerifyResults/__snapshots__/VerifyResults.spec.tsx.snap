// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerifyResults should render correctly 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-kpx2h6"
      data-testid="valid-box"
    >
      <h5
        aria-label="Document is Valid"
        class="MuiTypography-root MuiTypography-h5 css-156doar-MuiTypography-root"
        tabindex="0"
      >
        Valid
      </h5>
    </div>
    <div
      class="css-z38lsx-MuiStack-root"
    >
      <div
        class="MuiBox-root css-0"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          tabindex="0"
        >
          Issued by:
        </h6>
        <h6
          class="MuiTypography-root MuiTypography-h6 css-tjvwag-MuiTypography-root"
          style="word-wrap: break-word;"
          tabindex="0"
        >
          did:key:123
        </h6>
      </div>
      <div
        class="MuiBox-root css-fv3lde"
        data-testid="proof"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeMedium css-4b67si-MuiSvgIcon-root"
          data-testid="CheckCircleOutlineOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm4.59-12.42L10 14.17l-2.59-2.58L6 13l4 4 8-8z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-18x2lpd-MuiTypography-root"
          tabindex="0"
        >
          Document has not been tampered with
        </p>
      </div>
      <div
        class="MuiBox-root css-fv3lde"
        data-testid="status"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeMedium css-4b67si-MuiSvgIcon-root"
          data-testid="CheckCircleOutlineOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm4.59-12.42L10 14.17l-2.59-2.58L6 13l4 4 8-8z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-18x2lpd-MuiTypography-root"
          tabindex="0"
        >
          Document has not been revoked
        </p>
      </div>
    </div>
  </div>
</body>
`;
